src/MagnumPlugins/ObjImporter/ObjImporter.cpp

```diff
@@ -115,387 +372,575 @@ ObjImporter::~ObjImporter() = default;
-        } else if(keyword == "v") {
-            ++positionIndexOffset;
-            thisIsFirstMeshAndItHasNoData = false;
-        } else if(keyword == "vt") {
-            ++textureCoordinateIndexOffset;
-            thisIsFirstMeshAndItHasNoData = false;
-        } else if(keyword == "vn") {
-            ++normalIndexOffset;
-            thisIsFirstMeshAndItHasNoData = false;
-
-        /* Index data, just mark that we found something for first unnamed
-           object */
-        } else if(thisIsFirstMeshAndItHasNoData) for(const std::string& data: {"p", "l", "f"}) {
-            if(keyword == data) {
-                thisIsFirstMeshAndItHasNoData = false;
-                break;
-            }
+        /* Object name */
+        } else if(keyword == "o") {
+            finishSection();
+            finishObject();
+
+            ArrayView<char> name;
+            std::tie(name, pos) = nextWord(pos);
+
+            _state->objects.emplace_back(new ObjObject{name});
+            object = _state->objects.back().get();
+            group = nullptr;
+            meshData = nullptr;
+
+        /* Object group */
+        } else if(keyword == "g") {
+            //TODO: Handle geometry shared by multiple groups, e.g. g group1 group2
+            ensureObject();
+            object->groups.emplace_back(new ObjGroup{*object});
+            group = object->groups.back().get();
+
+            meshData = nullptr;
+
+            ArrayView<char> name;
+            std::tie(name, pos) = nextWord(pos);
+            group->name = std::string{name.data(), name.size()};
+
+        /* Load a material library */
+        } else if(keyword == "mtllib") {
+            ArrayView<char> word;
+            pos = skipWhitespaces(pos);
+            std::tie(word, pos) = nextWord(pos);
+
+            parseMaterialLibrary(word);
+
+        /* Set current material and add a new mesh for it */
+        } else if(keyword == "usemtl") {
+        ArrayView<char> word;
+        pos = skipWhitespaces(pos);
+            std::tie(word, pos) = nextWord(pos);
+
+            const int materialIndex = _state->materialIds[std::string{word.data(), word.size()}];
+            if(meshData == nullptr || materialIndex != meshData->materialId) {
+                /* Switching the material here, need to create a new mesh */
+                //TODO C++17
+                ensureGroup();
+                meshData = &group->meshDataForMaterial(materialIndex);
+            } // else: usemtl did not result in material switch, no need to create new mesh
+
+        /* Ignore unsupported keywords, error out on unknown keywords */
+        } else {
+            Warning() << "Trade::ObjImporter::parse(): unknown keyword:" << keyword;
+        }
+
+        if(sectionEnd) {
+            finishSection();
         }
 
         /* Ignore the rest of the line */
-        ignoreLine(*_file->in);
+        line = pos = skipWhitespaces(ignoreLine(pos));
```

Comment:
Ah. Right, confused it with this one.

So this means there can't be whitespace before the first keyword in the file but can be for all the other? Is that desired behavior? To make the code clearer, I would touch just a single line in every loop iteration, not going over -- so ignoring prefix whitespace at the beginning and ignoring suffix whitespace at the end.